### PR TITLE
AddIndexCommand - Command to send PropertyInfo index request to the MDAO

### PR DESCRIPTION
**Pull request by jlhughes**
Refactor EasyDAO, removing deprecated addIndex and provide addIndex support during delegate creation.

### CODE REVIEW

File: src/foam/dao/EasyDAO.js [hunk 1]
```diff
@@ -111,16 +115,25 @@ foam.CLASS({
...
+      Loggers.logger(getX(), this).warning("Of not found");
+      return "EasyDAO: DAO not found";
      `
     },
     {
       name: 'nSpec',
       class: 'FObjectProperty',
       type: 'foam.nanos.boot.NSpec',
-      javaPostSet: 'setName(val.getName());'
+      javaFactory: 'return getX().get(NSpec.class);',
```
Comment (kgrgreer) [inline src/foam/dao/EasyDAO.js:136]: What is this for?

File: src/foam/dao/EasyDAO.js [hunk 2]
```diff
@@ -356,6 +364,15 @@ foam.CLASS({
         if ( getPm() )
           delegate = new foam.dao.PMDAO.Builder(getX()).setNSpec(getNSpec()).setDelegate(delegate).build();
 
+        if ( indexes.size() > 0 ) {
+          Object result = delegate.cmd_(getX(), new AddIndexCommand(false, indexes.toArray(new PropertyInfo[indexes.size()])));
```
Comment (kgrgreer) [inline src/foam/dao/EasyDAO.js]: I think this will create one index with each component being what was intended to be a different index. One line 1131 below the PropertyInfo... is an array of parts of an index, but it looks like here each element is meant to be a different index.

File: src/foam/dao/EasyDAO.js [hunk 3]
```diff
@@ -1105,42 +1122,16 @@ model from which to test ServiceProvider ID (spid)`,
...
-      `
-    },
-
-    /** Only relevant if cache is true or if daoType
-      was set to MDAO, but harmless otherwise. Adds an existing index
-      to the MDAO.
-      @param index The index to add.
-    */
-    {
-      name: 'addIndex',
```
Comment (kgrgreer) [inline src/foam/dao/EasyDAO.js:1130]: Why is this removed?

File: src/foam/dao/EasyDAO.js [hunk 4]
```diff
@@ -1105,42 +1122,16 @@ model from which to test ServiceProvider ID (spid)`,
...
-      // TODO: The java Index interface conflicts with the js CLASS Index
-      args: [ { javaType: 'foam.dao.index.Index', name: 'index' } ],
-      code: function addIndex(index) {
-        this.mdao && this.mdao.addIndex.apply(this.mdao, arguments);
-        return this;
-      },
-      javaCode: `
-        DAO dao = (DAO) getMdao();
-        if ( dao != null && dao instanceof foam.dao.MDAO ) {
-          ((foam.dao.MDAO)dao).addIndex(index);
```
Comment (kgrgreer) [inline src/foam/dao/EasyDAO.js:1142]: This method is overloaded in Java with two versions: one that takes a PropertyInfo[] and one that takes an Index.

Comment (jlhughes) [reply src/foam/dao/EasyDAO.js:1142]: We don't use addIndex(Index) anywhere.

File: src/foam/dao/EasyDAO.js (hunk 3 above)
Comment (jlhughes) [reply src/foam/dao/EasyDAO.js:1130]: Deprecated and not used. We replaced them all with addPropertyIndex.

File: src/foam/dao/EasyDAO.js (hunk 2 above)
Comment (jlhughes) [reply src/foam/dao/EasyDAO.js]: This retains existing behaviour and 'protected' and used to capture indexes from Aware interfaces encountered during delegate creation, and adding them at then end of delegate creation.

File: src/foam/dao/EasyDAO.js (hunk 1 above)
Comment (jlhughes) [reply src/foam/dao/EasyDAO.js:136]: A few of the 'Aware' DAOs make use of the nspec name in their own setup.  The NSpecFactory places the NSpec in the calling/construction context.  EasyDAO is NSpecAware but this property is not always set before getName is called, so when not set, check the context for it.
